### PR TITLE
Replace RabbitMQ with Redis queues

### PR DESCRIPTION
**Pull request by mezis**
Using RabbitMQ in Routemaster adds unwarranted complexity, and does not significantly reduce latency or increase throughput.
It also makes for difficult testing as both RabbitMQ remote administration and the Ruby client (`bunny`) are fairly unstable.

This PR removes RabbitMQ and replaces it with Redis-backed queues.
Pushing, popping, (n)acking from queues is implemented with atomic Lua scripts.

The `INTERNALS.md` document was updated to detail the Redis schema in use.

There are strictly no API changes, and the integration tests were not modified here (apart from removing the per-spec RabbitMQ reset, of course).

I also used the opportunity to normalize Ruby namespacing, so probably best to read this with [`?w=1`](https://github.com/mezis/routemaster/pull/2/files?w=1).

### CODE REVIEW

File: INTERNALS.md [hunk 1]
```diff
@@ -27,10 +27,49 @@ This process is built from 4 key classes:
...
+
+`topics`
+
+  The set of topic names.
+
+`subscription`
+
+  The set of subscriber UUIDs.
+
+`topic/{name}`
```
Comment (gregbeech) [inline INTERNALS.md]: Would it be better to have `:` rather than `/` in the key name, which is more conventional?

File: routemaster/lua/nack.lua [hunk 2]
```diff
@@ -0,0 +1,12 @@
+--
+-- Negative-acknowledges a message.
+--
+-- KEYS[1]: list of new message UUIDs
+-- KEYS[2]: zset of pending (nacked) message UUIDs
```
Comment (gregbeech) [inline routemaster/lua/nack.lua:5]: Would it be worth putting the zset argument first for parity with the order of the arguments in the `ack` script? It's not really a big deal though.

Comment (gregbeech) [inline routemaster/lua/nack.lua:5]: Ah, maybe not, looking at the other scripts the arguments are all a bit different anyway, so it's probably fine.

File: routemaster/mixins/redis.rb [hunk 3]
```diff
@@ -2,13 +2,31 @@
...
+        sha = @@_lua_sha[script_name] and return sha
+
+        path = File.expand_path("../../lua/#{script_name}.lua", __FILE__)
+        sha = _redis.script('load', File.read(path))
+        @@_lua_sha[script_name] = sha
+      end
 
-  def _redis
-    @@_redis ||= Redis.new(url: ENV['ROUTEMASTER_REDIS_URL'])
+      def _lua_run(script_name, keys:nil, argv:nil)
```
Comment (gregbeech) [inline routemaster/mixins/redis.rb]: Would it make sense to name this something like `_redis_lua_run` as once mixed in the name `_lua_run` won't really give much context about _where_ the Lua is being run?

File: routemaster/models/event.rb [hunk 4]
```diff
@@ -3,47 +3,49 @@
...
+    class Event
+      include Mixins::Assert
+      extend  Mixins::Assert
+
+      VALID_TYPES = %w(create update delete noop)
+
+      attr_reader :topic, :type, :url, :timestamp
+
+      def initialize(topic:, type:, url:, timestamp: nil)
+        _assert VALID_TYPES.include?(type), 'bad event type'
```
Comment (gregbeech) [inline routemaster/models/event.rb:17]: This would fail with `type: :create`; would be nice to allow symbols as well as strings.

File: routemaster/models/queue.rb [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        "subscriber:#{@subscription.subscriber} id:0x#{object_id.to_s(16)}"
+      end
+
+      def inspect
+        "<#{self.class.name} #{self}>"
+      end
+
+      module ClassMethods
+        def push(subscriptions, message)
+          keys  = subscriptions.map { |sub|
```
Comment (gregbeech) [inline routemaster/models/queue.rb]: Could use `flat_map` here instead of `map.flatten`

File: INTERNALS.md (hunk 1 above)
Comment (gregbeech) [inline INTERNALS.md]: Also perhaps a `routemaster:` or even `rm:` prefix in case it's a shared Redis DB to avoid potential conflicts?

Comment (mezis) [inline INTERNALS.md]: Given this diff is a breaking change, might as well do both. Probably wouldn't want to share the Routemaster DB with anything else though ;)

File: routemaster/lua/nack.lua (hunk 2 above)
Comment (mezis) [inline routemaster/lua/nack.lua:5]: If that's confusing — I could pass the same 3 keys to all the scripts.
Do you think that'd help? The performance loss would be tiny.

File: routemaster/mixins/redis.rb (hunk 3 above)
Comment (mezis) [inline routemaster/mixins/redis.rb]: Yeah... I had the same thought when re-reading and I was lazy. Will fix.

File: routemaster/models/event.rb (hunk 4 above)
Comment (mezis) [inline routemaster/models/event.rb:17]: This is purely internal code so I'd rather err on the side of being strict. None of the app's code should use symbols here.

(I'd definitely agree in the _client_, though)

File: routemaster/models/queue.rb (hunk 5 above)
Comment (mezis) [inline routemaster/models/queue.rb]: Damn, always forget that one. Sure, will fix.

File: routemaster/lua/nack.lua (hunk 2 above)
Comment (gregbeech) [inline routemaster/lua/nack.lua:5]: I think it's OK tbh, it's only internal code really so as long as the right keys/args are being sent then it shouldn't be an issue.

**Comment by gregbeech** [comment]
Looks like a good change; Redis seems a pretty sensible way to back the queue infrastructure as pretty much everyone is running it now so should reduce the friction to adoption (including for us, as we need to scale Redis anyway, and we could entirely drop RabbitMQ).

I must admit that I don't really know the codebase so it's hard to follow such a large change and how it all fits together, but the atomic units look sensible. I guess the obvious concern with replacing the guts of something is it takes it back from 'production tested' to 'beta grade' almost, and so perhaps it would be better to keep Redis in a separate branch from `master` until it's had any issues ironed out?

**Comment by mezis** [comment]
> perhaps it would be better to keep Redis in a separate branch from master until it's had any issues ironed out?

Definitely makes sense to me. I'll do my best to bring in future improvements against the master branch, and backport them against a Redis branch.

(I'll take comments here, but I'll merge manually into a `beta-redis` branch)
